### PR TITLE
130344867: Remove the metrics data from metric collector db based on the cut off days

### PR DESCRIPTION
**Pull request by kanekoh**
[#130344867]

- Implemented prune functionality for metric collector - applicationmetrics table.
- Added corresponding tests

### CODE REVIEW

File: src/metricscollector/appmetricprune/prune/prune.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package prune
```
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune.go]: Why does this exist?

File: src/metricscollector/appmetricprune/prune/prune.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+func NewPrune(logger lager.Logger, metricDB db.MetricsDB) *Prune {
+	return &Prune{
+		logger:   logger,
+		metricDB: metricDB,
+	}
+}
+
+func (pr *Prune) PruneMetricsOlderThan(cutoffDays int) {
```
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune.go]: This is the only relevant code

File: src/metricscollector/appmetricprune/prune/prune.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+const TokenTypeBearer = "bearer"
+
+type Prune struct {
+	logger   lager.Logger
+	metricDB db.MetricsDB
+}
+
+type ErrorResponse struct {
```
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune.go]: Why is this here?

File: src/metricscollector/appmetricprune/prune/prune.go (hunk 1 above)
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune.go]: The naming isn't good. We don't need it to live so far down.
prune.go should just like in the metricscollector and call the struct Pruner.

You will then wire this in as an ifrit.Process that runs 1,2 or 4 times a day. I don't really care on the interval. We can make that configurable too.

File: src/metricscollector/appmetricprune/prune/prune_test.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		logger          lager.Logger
+		prune           *Prune
+		timeBeforePrune time.Time
+		timeAfterPrune  time.Time
+		cutoffDays      int
+	)
+
+	BeforeEach(func() {
+
+		logger = lager.NewLogger("prune-test")
```
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune_test.go]: Why does this need to be exposed? its only used when we create the Pruner below.

File: src/metricscollector/appmetricprune/prune/prune.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	return &Prune{
+		logger:   logger,
+		metricDB: metricDB,
+	}
+}
+
+func (pr *Prune) PruneMetricsOlderThan(cutoffDays int) {
+	timestamp := time.Now().AddDate(0, 0, -cutoffDays).UnixNano()
+
+	pr.metricDB.PruneMetrics(timestamp)
```
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune.go]: Error handling?

File: src/metricscollector/appmetricprune/prune/prune_test.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+		Context("prune metric data", func() {
+			BeforeEach(func() {
+				cutoffDays = 20
+
+				prune = NewPrune(logger, metricDB)
+			})
+
+			JustBeforeEach(func() {
+				timeBeforePrune = time.Now()
```
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune_test.go]: You can just do all of this in the It()

File: src/metricscollector/appmetricprune/prune/prune_test.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+				prune.PruneMetricsOlderThan(cutoffDays)
+				timeAfterPrune = time.Now()
+			})
+
+			It("It has the the correct timestamp cutoff for pruning", func() {
+				Expect(metricDB.PruneMetricsCallCount()).To(Equal(1))
+
+				Expect(metricDB.PruneMetricsArgsForCall(0) >= timeBeforePrune.AddDate(0, 0, -cutoffDays).UnixNano()).To(Equal(true))
```
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune_test.go]: Just use BeTemporally

File: src/metricscollector/config/config.go [hunk 8]
```diff
@@ -54,12 +55,21 @@ var defaultCollectorConfig = CollectorConfig{
 	PollInterval:    DefaultPollInterval,
 }
 
+type DataPruneConfig struct {
+	CutoffDays int `yaml:"cutoff_days"`
+}
+
+var defaultDataPruneConig = DataPruneConfig{
```
Comment (fraenkel) [inline src/metricscollector/config/config.go]: Just so you know, I am not convinced I want the pruning to be done by the metricscollector.

I think we will have a standalone pruner that will do it for all the DBs. So this code should just move to src/autoscaler/pruner as the top level package. The pruners themselves can live at the root and the main would be autoscaler/pruner/cmd/pruner.

Comment (ghaih) [inline src/metricscollector/config/config.go]: Just thinking aloud for clarity,

We no longer want to have any pruning code in respective packages i.e in metricscollector or data aggregator etc

The approach is to have 

- autoscaler/src/pruner/cmd/ pruner.go -> running the specific ifrit pruner processesas follows
 -  autoscaler/src/pruner/metriccollector_metric_pruner.go
 - autoscaler/src/pruner/aggregator_metric_pruner.go etc

Comment (fraenkel) [inline src/metricscollector/config/config.go]: you are duplicating yourself with your naming. You are already in the pruner package so metriccollector or whatever db table name .go is sufficient, pruner is redundant.
And these things are doing nothing more than driving the underlying DB code or in the case of scheduler, making the remote call.

Comment (ghaih) [inline src/metricscollector/config/config.go]: Yeah, those were not intended names, just to put across our understanding for clarity to depict their jobs used those names:)

So we will go ahead with this structure which is more clear. We did struggle with the previous approach.

File: src/metricscollector/appmetricprune/prune/prune_test.go (hunk 7 above)
Comment (ghaih) [inline src/metricscollector/appmetricprune/prune/prune_test.go]: It is not time but int64 so will change to use BeNumerically?

File: src/metricscollector/appmetricprune/prune/prune.go (hunk 2 above)
Comment (ghaih) [inline src/metricscollector/appmetricprune/prune/prune.go]: You mean have the cutoffDays in the struct Pruner

File: src/metricscollector/appmetricprune/prune/prune_test.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+	BeforeEach(func() {
+
+		logger = lager.NewLogger("prune-test")
+
+		metricDB = &fakes.FakeMetricsDB{}
+
+	})
+	Describe("when a request for prune old metric data comes", func() {
+
```
Comment (ghaih) [inline src/metricscollector/appmetricprune/prune/prune_test.go]: Didn't get it. Check logs for any errors?

File: src/metricscollector/appmetricprune/prune/prune.go (hunk 2 above)
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune.go]: yes, it doesn't change. You need to think about how this is really going to be used. I envision you create a Pruner that is just called on some time sequence. The cutoff is always the same.

File: src/metricscollector/appmetricprune/prune/prune_test.go (hunk 9 above)
Comment (fraenkel) [inline src/metricscollector/appmetricprune/prune/prune_test.go]: Where are the tests when errors occur?

File: src/autoscaler/pruner/cmd/pruner/pruner_test.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gbytes"
+	. "github.com/onsi/gomega/gexec"
+	"io/ioutil"
+	"os"
+)
+
+var _ = Describe("Pruner", func() {
```
Comment (fraenkel) [inline src/autoscaler/pruner/cmd/pruner/pruner_test.go]: We actually don't verify that the MetricsPruner starts.

File: src/autoscaler/pruner/config/config.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	Level: DefaultLoggingLevel,
+}
+
+type DbConfig struct {
+	MetricsDbUrl string `yaml:"metrics_db_url"`
+}
+
+type PrunerConfig struct {
+	IntervalInHours int `yaml:"interval_in_hours"`
+	CutoffDays      int `yaml:"cutoff_days"`
```
Comment (fraenkel) [inline src/autoscaler/pruner/config/config.go]: I have a feeling that the cutoff will be different per DB but we can leave this here and fix it once we add a second DB/Table.

File: src/autoscaler/pruner/metricsdb.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+		metricsDB:  metricsDB,
+		interval:   time.Duration(interval) * time.Hour,
+		cutoffDays: cutoffDays,
+		clock:      clock,
+		doneChan:   make(chan bool),
+	}
+}
+
+func (mdp *MetricsDBPruner) Start() {
+	mdp.ticker = mdp.clock.NewTicker(mdp.interval)
```
Comment (fraenkel) [inline src/autoscaler/pruner/metricsdb.go]: why does the ticker live here? It should be scoped to the startMetricPrune which is really pruneMetrics().

File: src/autoscaler/pruner/metricsdb_test.go [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+		It("Stop the pruner", func() {
+			fclock.Increment(TestRefreshInterval)
+			Eventually(metricsDB.PruneMetricsCallCount).Should(Equal(2))
+
+			pruner.Stop()
+			Eventually(buffer).Should(gbytes.Say("metrics-db-pruner-stopped"))
+
+			fclock.Increment(TestRefreshInterval)
+			Eventually(metricsDB.PruneMetricsCallCount).Should(Equal(2))
```
Comment (fraenkel) [inline src/autoscaler/pruner/metricsdb_test.go]: You mean Consistently.

File: src/autoscaler/pruner/config/config.go (hunk 11 above)
Comment (ghaih) [inline src/autoscaler/pruner/config/config.go]: Yes thought about it. Will change as part of data aggregator prune story.

File: src/autoscaler/pruner/cmd/pruner/pruner_test.go (hunk 10 above)
Comment (ghaih) [inline src/autoscaler/pruner/cmd/pruner/pruner_test.go]: Will add the check in prune_suite_test and a another in pruner_test.

File: src/autoscaler/pruner/metricsdb_test.go (hunk 13 above)
Comment (ghaih) [inline src/autoscaler/pruner/metricsdb_test.go]: Yes.

File: src/autoscaler/pruner/metricsdb.go (hunk 12 above)
Comment (ghaih) [inline src/autoscaler/pruner/metricsdb.go]: Moving it to startMetricPrune results in 2 data race. So will have it in Start.

**Comment by fraenkel** [comment]
Merged with some minor changes
